sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java

```diff
@@ -298,4 +298,38 @@ public void run() {
         }).start();
         mAuthTokenContext.setAuthToken(AUTH_TOKEN, "some-id", null);
     }
+
+    @Test(timeout = 5000)
+    public void listenerDeadlockCheckIfTokenNeedsToBeRefreshed() {
+        final CountDownLatch latch1 = new CountDownLatch(1);
+        final CountDownLatch latch2 = new CountDownLatch(1);
+        mAuthTokenContext.addListener(new AbstractTokenContextListener() {
+            @Override
+            public void onTokenRequiresRefresh(String homeAccountId) {
+                latch1.countDown();
+                try {
+                    latch2.await();
+                } catch (InterruptedException ignored) {
+                }
+            }
+        });
+        new Thread(new Runnable() {
+
+            @Override
+            public void run() {
+
+                /* Wait for listener call. */
+                try {
+                    latch1.await();
+                } catch (InterruptedException ignored) {
+                }
+
+                /* Call something synchronized. */
+                assertEquals(AUTH_TOKEN, mAuthTokenContext.getAuthToken());
+                latch2.countDown();
+            }
+        }).start();
+        AuthTokenInfo info =  new AuthTokenInfo(AUTH_TOKEN, mock(Date.class),  mock(Date.class));
```

Comment:
Please format. Run the tool for it. There is a wiki page with detailed desc for it.